src/lib/comms/sol-bluetooth-impl-linux.c

```diff
@@ -0,0 +1,683 @@
+    SOL_NULL_CHECK_GOTO(d, skip);
+
+    info = &d->info;
+
+    if (info->name)
+        goto skip;
+
+    r = sd_bus_message_read_basic(m, 's', &name);
+    SOL_INT_CHECK_GOTO(r, < 0, skip);
+
+    /* When we receive the property again we will try to allocate again */
+    info->name = strdup(name);
+
+    return true;
+
+skip:
+    r = sd_bus_message_skip(m, "s");
+    SOL_INT_CHECK(r, < 0, false);
+
+    return false;
+}
+
+static bool
+device_property_uuids_set(void *data, const char *path, sd_bus_message *m)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+    struct sol_bt_device_info *info;
+    struct sol_bt_uuid *u = NULL;
+    char **uuids = NULL, **uuid;
+    int r;
+
+    d = find_device_by_path(ctx, path);
+    SOL_NULL_CHECK_GOTO(d, skip);
+
+    info = &d->info;
+
+    r = sd_bus_message_read_strv(m, &uuids);
+    SOL_INT_CHECK_GOTO(r, < 0, skip);
+
+    sol_vector_clear(&info->uuids);
+    for (uuid = uuids; uuid && *uuid; uuid++) {
+        u = sol_vector_append(&info->uuids);
+        SOL_NULL_CHECK_GOTO(u, done);
+
+        r = sol_bt_uuid_from_str(u, *uuid);
+        SOL_INT_CHECK_GOTO(r, < 0, done);
+    }
+
+done:
+    if (!u || r)
+        sol_vector_clear(&info->uuids);
+
+    if (uuids)
+        strv_free(uuids);
+
+    return true;
+
+skip:
+    r = sd_bus_message_skip(m, "as");
+    SOL_INT_CHECK(r, < 0, false);
+
+    return false;
+}
+
+static bool
+device_property_rssi_set(void *data, const char *path, sd_bus_message *m)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+    struct sol_bt_device_info *info;
+    int16_t rssi;
+    int r;
+
+    d = find_device_by_path(ctx, path);
+    SOL_NULL_CHECK(d, false);
+
+    info = &d->info;
+
+    r = sd_bus_message_read_basic(m, 'n', &rssi);
+    SOL_INT_CHECK(r, < 0, false);
+
+    info->rssi = rssi;
+
+    return true;
+}
+
+static const struct sol_bus_properties device_properties[] = {
+    [DEVICE_PROPERTY_ADDRESS] = {
+        .member = "Address",
+        .set = device_property_address_set,
+    },
+    [DEVICE_PROPERTY_NAME] = {
+        .member = "Name",
+        .set = device_property_name_set,
+    },
+    [DEVICE_PROPERTY_UUIDS] = {
+        .member = "UUIDs",
+        .set = device_property_uuids_set,
+    },
+    [DEVICE_PROPERTY_RSSI] = {
+        .member = "RSSI",
+        .set = device_property_rssi_set,
+    },
+    { NULL, NULL }
+};
+
+static void
+device_property_changed(void *data, const char *path, uint64_t mask)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+    struct sol_bt_device_info *info;
+    const uint64_t all_present = (1 << DEVICE_PROPERTY_ADDRESS) |
+        (1 << DEVICE_PROPERTY_NAME) |
+        (1 << DEVICE_PROPERTY_RSSI);
+    struct sol_bt_scan_pending *scan = &ctx->pending_scan;
+
+    d = find_device_by_path(ctx, path);
+    SOL_NULL_CHECK(d);
+
+    d->mask |= mask;
+
+    if (!scan->callback)
+        return;
+
+    info = &d->info;
+
+    if ((d->mask & all_present) == all_present)
```

Comment:
why only on all are present? Wouldn't be better to use a single `sol_timeout_add(0,...)` to dispatch it? so you delay it to the next mainloop iteration, but do it even if not all properties arrived?